File: 04-Giant-Squid.js
const readline = require('readline')
const rl = readline.createInterface({ input: process.stdin })
const boards = []
let drawSequence

rl.on("line", data => {
  if (data === '')
    boards.push([])
  else
    if (data.includes(' '))
      boards[boards.length-1].push(data.trim('').split(/ +/).map(v=>Number.parseInt(v)))
    else 
    drawSequence = data.split(',').map(v=>Number.parseInt(v))
})

rl.on('close', () => {

  const result = play(drawSequence, boards)
  console.log('partOne', partOne(result))
  console.log('partTwo', partTwo(result))
})


function play(drawSequence, boards) {
  return boards.map(playOne)

  function playOne(board) {
    function mark(draw) {
      for (let row of board) {
        const idx = row.indexOf(draw)
        if (idx !== -1)
          row[idx] = 'X'
      }
    }

    const isAWin = nums => nums.reduce((w, c) => w&& c==='X', true)
    function isWinningBoard() {
      for (const row of board) {
        if (isAWin(row)) return true
      }
      for (let i=0; i<board.length;i+=1) {
        const column = board.map(row => row[i])
        if (isAWin(column)) 
          return true
      }
      return false
    }

    for (let numDraws=0; numDraws < drawSequence.length; numDraws+=1) {
      const draw = drawSequence[numDraws]
      mark(draw)
      if (isWinningBoard()) {
        return { board, numDraws, draw }
      }
    }
  }
}

const score = winning => winning.board.flatMap(_=>_).filter(c=>c!=='X').reduce((a,b)=>a+b) * winning.draw

function partOne(result) {
  const winning = result.reduce(
    (winning, current) => winning.numDraws < current.numDraws ? winning : current,
    {numDraws: Number.MAX_SAFE_INTEGER}
  )

  return score(winning)
}

function partTwo(result) {

  const winning = result.reduce(
    (winning, current) => winning.numDraws > current.numDraws ? winning : current,
    {numDraws: 0}
  )

  return score(winning)
}

File: 04-Giant-Squid.hs
module GiantSquid where
-- runhaskell 04-Giant-Squid.hs<04-input.txt

import Data.List.Split (splitOn)
import Data.List (isSubsequenceOf, sort, transpose,sortOn)
import Utils (atoi, atoiList, splitlist, flatten)

type Draws = [Int]
type Board = [[Int]]

-- | map a sentence to a list of numbers
toiList :: String -> [Int]
toiList = map atoi . words

-- | parse a list of strings to a list of bingo boards
parseBoards :: [String] -> [Board]
parseBoards = splitlist 5 . map toiList

-- | parse the lines of the file into
-- | - a list of numbers to be called/drawn
-- | - a list of bingo boards
parseInput :: [String] -> (Draws, [Board])
parseInput input = (drawSequence, boards) 
  where
    drawSequence = atoiList $ head input
    boards = parseBoards $ filter (/="") $ tail input

-- | A board is a winning board if 
-- | all numbers in a row or column have been drawn
isWinningBoard :: Draws -> Board -> Bool
isWinningBoard draws board = winningRows || winningColumns
  where 
    sortedDraws = sort draws
    winningRows = any isWin board 
    winningColumns = any isWin $ transpose board
    isWin nums = isSubsequenceOf (sort nums) sortedDraws

-- | Given all draws and a board
-- | calculate the number of draws needed to win
play :: Draws -> Board -> (Board, Int)
play draws board = (board, numDraws)
   where 
     (numDraws, _) = foldl step (0, False) [1..(length draws)]
     step (numDraws, True)  _ = (numDraws, True)
     step (_,        False) n = (n, (isWinningBoard (take n draws) board))

-- | The score of the winning board is
-- | the sum of all unmarked numbers on that board
-- | multiply that sum by the number that was just called when the board won
score :: [Int] -> (Board, Int) -> Int
score draws (board, numDraws) = boardScore * winningDraw
  where 
    boardScore = foldl (+) 0 unmarkedNumbers 
    winningDraw = draws !! (numDraws-1)
    drawsToWin = take numDraws draws
    unmarkedNumbers = filter (\i -> notElem i drawsToWin) $ flatten board

-- | Calculate the number of draws needed to win for all board
-- | sort ascending by number of draws
playAll :: Draws -> [Board] -> [(Board, Int)]
playAll draws boards = sortOn snd $ map (play draws) boards

main = do
  contents <- getContents
  let input = lines contents
      (drawSequence, boards) = parseInput input
      result = playAll drawSequence boards
     
  print $ score drawSequence $ head result
  print $ score drawSequence $ last result
